### DISCUSSION TITLE
LAB 03.01 Error

### DISCUSSION THREAD

**Discussion by Jucamilocar08** [rramosp/ai4eng.v1#276] [03 - GENERACION DE MODELOS]
Hola profe, es que tengo el siguiente error, al ejecutar mi codigo en el laboratorio 2 del modulo 3.

este es mi codigo

     import itertools

     import numpy as np

    def fit(X,y):

    x = np.linspace(0, 1, num=11, retstep=True,)[0]
    t = np.array(list(itertools.product(x, repeat = 2)))

    h = []

    for i in range(len(t)):
        s = predict(X,t[i])
        h.append(np.mean(y==s))

    t = np.array(t[np.argmax(h)])

    return np.array(np.around(t,2))

y este es el error que me sale, despues de hacer el login.

error: ERROR EXECUTING YOUR CODE

Traceback (most recent call last):
  File "/content/calling_script.py", line 24, in <module>
    r = eval("tkgbypyi.{}(source_functions, source_variables, '[juan.cardona50@udea.edu.co](mailto:juan.cardona50@udea.edu.co)')".format(grader_function_name), namespace)
  File "<string>", line 1, in <module>
  File "/content/tkgbypyi.py", line 32, in grader_02
    ts = st_fit(X, y)
  File "<string>", line 9, in fit
NameError: name 'predict' is not defined
 (set session.debug=True for tracebacks)

Creo que error se da por que, el corrector no ve la funcion predict que hice en el ejercicio anterior, y es por eso que se produce el error.

**Comment by Felipe-RA** [comment]
Saludos @Jucamilocar08 

En sesión sincronica miramos tu caso, pero igual para que quede por aquí el registro y le pueda ser util a alguien más, te comento:

Como bien dices, este error ocurre porque la función `predict` no se encuentra dentro de lo que el corrector puede ""ver""

Este concepto lo conocemos como `espacio de nombre en ambiente de ejecución` (execution environment namespace), un concepto estrechamente relacionado al `scope` o `alcance` de una función o variable, pero con diferencias puntuales significativas.

Un `espacio de nombre` (namespace), es la forma en la que Python gestiona las variables, funciones y clases. Puedes pensarlo como la cajita que guarda nombres, y la asociación de esos nombres a objetos.

Cada ambiente de ejecución tiene un `namespace` distinto.

Ahora, en este curso particularmente se manejan dos ambientes de ejecución. El de tu notebook (que esta alojado en tu pc o en los servers de Google Colab) , y el del corrector (que esta alojado en el servidor del curso)

Puedes entenderlo de mejor forma con el siguiente gráfico que busca lo que pasa cuando le das al `submit_task`

![image](https://github.com/rramosp/ai4eng.v1/assets/49454068/e03bb9b9-9fae-4670-a050-853cb5489373)


Cuando corres `submit_task`  lo que se hace es que se envía variables y/o funciones con un determinado nombre, así como **todo lo que esté contenido dentro de la función enviada**; y se incorpora al namespace siguiendo una estrategia llamada LEGB.
En tu caso el grader recibió a través del `submit_task` sólo tu función `fit()`, pero no recibió `predict` pues no estaba contenida dentro de la función `fit()`, como se pedia en el ejercicio.

Tu función `predict` nunca entró al `namespace` del ambiente de ejecución del corrector, es por esto el fallo en el grader.

Y en tu notebook si la encuentra puesto que si existe dentro del namespace de su ambiente de ejecución.

Gracias por publicar tu duda, espero que esto pueda servirle a alguien más!
